test/triangular_test.js

```diff
@@ -0,0 +1,255 @@
+          data: [
+            0, 0, 0,
+            0, 0, 0,
+            0, 0, 0,
+          ],
+        },
+        {
+          diagonal: 4,
+        },
+    );
+  });
+
+  it('triangular diagonal=-1', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            0, 6, 3,
+          ],
+        },
+        {
+          diagonal: -1,
+        },
+    );
+  });
+
+  it('triangular fully copied diagonal=-4', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          diagonal: -4,
+        },
+    );
+  });
+
+  it('triangular upper=false', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 0, 0,
+            9, 4, 0,
+            2, 6, 3,
+          ],
+        },
+        {
+          upper: false,
+        },
+    );
+  });
+
+  it('triangular upper=false diagonal=1', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 0,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          upper: false,
+          diagonal: 1,
+        },
+    );
+  });
+
+  it('triangular fully copied upper=false diagonal=4 ', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          upper: false,
+          diagonal: 4,
+        },
+    );
+  });
+
+  it('triangular upper=false diagonal=-1', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            0, 0, 0,
+            9, 0, 0,
+            2, 6, 0,
+          ],
+        },
+        {
+          upper: false,
+          diagonal: -1,
+        },
+    );
+  });
+
+  it('triangular fully zero upper=false diagonal=-4', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
```

Comment:
All the shapes are square. We should include at least one asymmetric case (different height and width) to ensure we don't have any bugs where we accidentally swap rows for columns (e.g. `shape: [4, 3]`.)

Reply:
Thanks @fdwr .
Last I referred to https://en.wikipedia.org/wiki/Triangular_matrix:
```
In mathematics, a triangular matrix is a special kind of [square matrix](https://en.wikipedia.org/wiki/Square_matrix).
``
OK, I will update it without restricting square matrix.

Reply:
I've read other definitions like "An m×n matrix A is called upper triangular if all entries lying below the diagonal entries are zero" which imply m and n can be distinct, but it's certainly that that square matrices (a subset) satisfy that too. ONNX doesn't care. Even Wolfram Alpha accepts the following fine `MatrixForm[UpperTriangularize[{{1, 2, 3, 4}, {4, 5, 6, 7}, {7, 8, 9, 10}}]]`.

> OK, I will update it without restricting square matrix.

👍